optaplanner-examples/src/main/java/org/optaplanner/examples/examination/solver/score/ExaminationConstraintProvider.java

```diff
@@ -0,0 +1,241 @@
+                        (topicConflict, leftExam, config) -> topicConflict.getStudentSize() * config.getTwoInARowPenalty());
+    }
+
+    protected Constraint twoExamsInADay(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(TopicConflict.class)
+                .join(Exam.class,
+                        equal(TopicConflict::getLeftTopic, Exam::getTopic),
+                        filtering((topicConflict, leftExam) -> leftExam.getPeriod() != null))
+                .ifExists(Exam.class,
+                        equal((topicConflict, leftExam) -> topicConflict.getRightTopic(), Exam::getTopic),
+                        equal((topicConflict, leftExam) -> leftExam.getDayIndex(), Exam::getDayIndex),
+                        // Find exams in a day, but not being held right after each other. That case is handled in the twoExamsInARow constraint.
+                        filtering((topicConflict, leftExam,
+                                rightExam) -> getPeriodIndexDifferenceBetweenExams(leftExam, rightExam) > 1))
+                .join(ExaminationConstraintConfiguration.class)
+                .penalize("Two exams in a day",
+                        HardSoftScore.ONE_SOFT,
+                        (topicConflict, leftExam, config) -> topicConflict.getStudentSize() * config.getTwoInADayPenalty());
+    }
+
+    protected Constraint periodSpread(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(ExaminationConstraintConfiguration.class)
+                .join(TopicConflict.class)
+                .join(Exam.class,
+                        equal((config, topicConflict) -> topicConflict.getLeftTopic(), Exam::getTopic),
+                        filtering((config, topicConflict, leftExam) -> leftExam.getPeriod() != null))
+                .ifExists(Exam.class,
+                        equal((config, topicConflict, leftExam) -> topicConflict.getRightTopic(), Exam::getTopic),
+                        filtering((config, topicConflict, leftExam, rightExam) -> rightExam.getPeriod() != null),
+                        filtering((config, topicConflict, leftExam,
+                                rightExam) -> getPeriodIndexDifferenceBetweenExams(leftExam,
+                                        rightExam) < (config.getPeriodSpreadLength() + 1)))
+                .penalize("Period spread",
+                        HardSoftScore.ONE_SOFT,
+                        (config, topicConflict, leftExam) -> topicConflict.getStudentSize());
+    }
+
+    protected Constraint mixedDurations(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(Exam.class)
+                .ifExists(Exam.class, equal(Exam::getPeriod, Exam::getPeriod))
+                .ifNotExistsOther(Exam.class,
+                        equal(Exam::getPeriod, Exam::getPeriod),
+                        equal(Exam::getRoom, Exam::getRoom),
+                        lessThan(Exam::getId, Exam::getId))
```

Comment:
return constraintFactory.from(Exam.class)
                .ifNotExists(Exam.class,
                        equal(Exam::getPeriod),
                        equal(Exam::getRoom),
                        lessThan(Exam::getId))

But that won't change much, if anything.

Reply:
Also, I'd put the `lessThan` filter first. Will eliminate many thing quickly.